modules/rostests/apitests/kernel32/UEFIFirmware.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * PROJECT:     ReactOS API Tests
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Tests for UEFI Firmware functions
+ * COPYRIGHT:   Copyright 2023 Ratin Gao <ratin@knsoft.org>
+ */
+
+#include "precomp.h"
+
+#include <ndk/psfuncs.h>
+#include <ndk/setypes.h>
+#include <ndk/sefuncs.h>
+#include <ndk/obfuncs.h>
+
+#define _A2W(quote) __A2W(quote)
+#define __A2W(quote) L##quote
+
+#define EFI_TEST_GUID_STRING "{8768B7AC-F82F-4120-B093-30DFA27DA3B5}"
+#define EFI_TEST_VARIABLE_NAME "RosUefiVarTest"
+
+#define EFI_DUMMY_NAMESPACE_GUID_STRING "{00000000-0000-0000-0000-000000000000}"
+#define EFI_DUMMY_VARIABLE_NAME ""
+
+static ULONG RandomSeed;
+static DWORD EfiVariableValue;
+
+static VOID test_GetFirmwareType(BOOL bIsUEFI)
+{
+#if (_WIN32_WINNT >= 0x0602)
+    BOOL bResult;
+    FIRMWARE_TYPE FirmwareType = FirmwareTypeUnknown, FirmwareExpect;
+
+    /* Test GetFirmwareType, should return FirmwareTypeBios or FirmwareTypeUefi */
+    bResult = GetFirmwareType(&FirmwareType);
+
+    ok(bResult,
+       "GetFirmwareType failed with error: 0x%08lX\n",
+       GetLastError());
+
+    if (!bResult)
+        return;
+
+    FirmwareExpect = (bIsUEFI ? FirmwareTypeUefi : FirmwareTypeBios);
+    ok(FirmwareType == FirmwareExpect,
+       "FirmwareType is %d, but %d is expected.\n",
+       FirmwareType, FirmwareExpect);
+#else
+    skip("This test can be ran only when compiled for NT >= 6.2.\n");
```

Comment:
```suggestion
    skip("This test can be run only when compiled for NT >= 6.2.\n");
```

Reply:
Note that alternatively, you could try loading at runtime the GetFirmwareType function with LoadLibrary("kernel32") + GetProcAddress("GetFirmwareType") and work from there.

Reply:
If that is possible it would be preferred, because that means the test can also run if ReactOS is not NT >= 6.2, but has the function available anyway.